### DISCUSSION TITLE
Native AOT support for DynamoDB

### DISCUSSION THREAD

**Discussion by mrdaghestani** [aws/aws-sdk-net#3194] [General]
Hi,
I understand that NativeAOT is still a very new feature and probably most people prefer not to use it but, I wanted to suggest that we can talk about supporting it in DynamoDB SDK.

I was thinking it could be very useful if there was a source generator for mapping DynamoDB items into database models inside our code base so we can ignore using reflection and probably that would solve the support for NativeAOT.

What do you think? Is it something that you're thinking of implementing? If yes, what's the schedule?

Thank you

**Comment by alexander-miglis** [comment]
I agree, I would really want this feature to be able to reduce lambda startup time.

**Comment by hans-voralberg** [comment]
It's 2024 and still DynamoDB high level API doesn't support .NET AOT.
Can you please tell us if you are working on it or if you plan to do it?

It's fundamental to know that for a lot of companies/devs. For example, because of this, we don't know if keeping using .NET AOT or not.

**Comment by normj** [comment]
Getting the DynamoDB high level libraries to support Native AOT trimming is a large effort. We would like to do it but between all of our other work to keep up with AWS we haven't started this effort. I do appreciate the upvotes and letting us know how important this feature is for you all. It really helps with the argument to prioritize the work.

**Comment by jamiewinder** [comment]
Thanks @normj. Based on my limited use of DocumentModel, it doesn't 'feel' very reflection-heavy. Do we know if it's viable to just disable trimming for this library for now?

**Comment by normj** [comment]
@jamiewinder That is good to know you are using DocumentModel. The DataModel that uses POCOs is where a lot of the work is. Maybe we could try and do the DocumentModel first.

To get around this can add a trimming descriptor file to your project to tell it to preserve all in the DynamoDB assembly. If you are using the DataModel you also need to add the assembly that has the POCOs uses by the DataModel. https://github.com/dotnet/runtime/blob/main/docs/tools/illink/data-formats.md#descriptor-format

**Reply by hans-voralberg** [reply]
@normj thanks for the clarifications.
According to this, https://github.com/aws/aws-sdk-net/issues/2542 , and to what I've personally tried, it's possible to make it work using some directives. 
But:
- That's not the way .NET AOT is intended to be used
- It's very break-risky and a lot of testing is needed
- Theoretically, by doing that, should it slow down the cold start time whe used in lambda functions?

Regarding lambdas:
AWS is telling us "_Finally you can use .NET with lambda without suffering from cold starts problems anymore because you can use AOT!_", but AOT is not usable with a so important service like DynamoDb...

Another solution could be to use low-level API, which is ok with AOT, but this complicates a lot all the work to do.

So finally, yes, I can say too, that if you can please give us an update for at least DocumentModel first, it will be great. And subsequently to plan it also for DataModel.

**Reply by mrdaghestani** [reply]
Thank you both, and I agree with @hans-voralberg.

Initially, I tried using DocumentModel, but it required a significant amount of time and effort, especially for developers accustomed to working with ORMs. Additionally, I attempted to design a BaseRepository to handle the mapping, but it still required manual mapping to be compatible with AOT compilation.

Finally, I considered writing a SourceGenerator, similar to what is available for [JSON serialization](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?pivots=dotnet-8-0). However, I believe that if the AWS SDK could provide such a feature, it would be beneficial for everyone.

**Comment by IvanFarkas** [comment]
@normj Where do I find a .NET 8, .NET 9 SAM Lambda AOT roadmap, please?

We are planning for our **Digital Agent Speech Engine** and need to know what directions to take and what will be available in the next 6, 12 month.
It has been implemented in .NET 6 SAM Lambda WebAPI.
We would like to move all or some pieces to .NET 8, .NET 9 SAM Lambda AOT.
Can you provide us some roadmap, please?

**Technologies used:**

- DynamoDB (**DataModel** preferred)
- S3
- Polly
- Translate
- Systems Management
- Cognito (or some other form of JWT Auth)

**Comment by IvanFarkas** [comment]
@normj Have a great Labor Day weekend.

**Comment by petro2050** [comment]
I'm too interested in AOT support. Seems like DynamoDB is my only blocker. +1 for having a roadmap. :)

**Comment by normj** [comment]
We haven't started AOT work on the DynamoDB high level libraries like the DataModel. If you want to use AOT with DynamoDB you need to use the low level service client. We are deep in getting the [V4 version](https://github.com/aws/aws-sdk-net/issues/3362) of the SDK out along with the usual keeping up with the AWS services. I would love to be able to tell you when we can get to the DynamoDB high level libraries but our priorities are always a delicate balance of what we want to do for the .NET SDK platform and supporting all the upcoming service changes. I do appreciate the +1s as it does help prioritize when we get a breather supporting AWS service launches.

@IvanFarkas To your other questions the .NET SDK supports all of the services you mentioned except for Translate streaming, not sure if that was what you meant by Translate. We will always continue to support new versions of .NET. Lambda managed runtimes follow the LTS cycle so no .NET 9 managed runtime since that is not an LTS. Next managed runtime would be .NET 10 unless something unusual happens. But if you are doing AOT the managed runtime doesn't matter because you are creating a standalone executable when you deploy to Lambda. Hope you had a great Labor Day weekend 😄 .

**Reply by IvanFarkas** [reply]
Thank you @normj for the detailed answer.

**Comment by normj** [comment]
Closing the issue now that Native AOT support as been released as part of V4 preview 6. Follow the this GitHub issue for tracking the status of V4: https://github.com/aws/aws-sdk-net/issues/3362
